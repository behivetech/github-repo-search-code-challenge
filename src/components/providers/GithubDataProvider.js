import React, {createContext, useContext, useReducer} from 'react';
import PropTypes from 'prop-types';
import useSWR from 'swr';

import fetcher from 'tools/fetcher';
import {useNotifications} from 'components/providers/NotificationProvider';

const getAppSessionState = () => {
    const appSessionState =
        typeof sessionStorage !== 'undefined' ? sessionStorage.getItem('appState') : null;

    return appSessionState ? JSON.parse(appSessionState) : {};
};
const INITIAL_STATE = {
    q: '',
    sort: '',
    order: 'desc',
    ...getAppSessionState(),
};
const DEFAULT_CONTEXT = {
    ...INITIAL_STATE,
    search: () => null,
};

const GithubDataContext = createContext(DEFAULT_CONTEXT);

export function useGithubData() {
    return useContext(GithubDataContext);
}

function reducer(state, {payload, type}) {
    const actions = {
        SET_SEARCH: {...state, ...payload},
    };

    const newState = {...state, ...actions[type]} || state;

    sessionStorage.setItem('appState', JSON.stringify(newState));

    return newState;
}

export default function GithubData({children}) {
    const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
    const {setNotification} = useNotifications();
    const {data, isValidating} = useSWR(
        ['/search/repositories', state],
        (path) => fetcher(path, state),
        {
            onError: (error) => {
                setNotification({
                    devMessage: error,
                    message: `There was a problem fetching the data. ${error}`,
                    messageKey: 'fetchDataError',
                    ttl: 20000,
                });
            },
            onSuccess: (response) => {
                if (response && response.message) {
                    setNotification({
                        message: response.message,
                        messageKey: 'fetchMessage',
                        ttl: 20000,
                    });
                }
            },
        }
    );

    function search(payload) {
        dispatch({type: 'SET_SEARCH', payload});
    }

    const context = {
        ...state,
        data,
        loading: isValidating,
        search,
    };

    return (
        <GithubDataContext.Provider value={context}>
            {children}
        </GithubDataContext.Provider>
    );
}

GithubData.propTypes = {
    children: PropTypes.node,
};
