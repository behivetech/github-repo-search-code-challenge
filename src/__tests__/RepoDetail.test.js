jest.mock('components/providers/GithubDataProvider');
import React from 'react';
import {render} from '@testing-library/react';
import * as nextRouter from 'next/router';
import RepoDetailPage from 'pages/repo/[repoId]';
import {useGithubData} from 'components/providers/GithubDataProvider';

describe('RepoDetailPage', () => {
    const mockData = {
        items: [
            {
                created_at: '2000-01-01T12:00:00Z',
                description: 'mock_description',
                forks_count: 123,
                name: 'mock_name',
                id: 1234,
                language: 'mock_language',
                license: {name: 'mock_name', url: 'mock_url'},
                open_issues: 5,
                owner: {
                    avatar_url: 'mock_avatar_url',
                    login: 'mock_login',
                    url: 'mock_url',
                },
                pushed_at: '2000-01-01T12:00:00Z',
                stargazers_count: 123456,
                updated_at: '2000-01-01T12:00:00Z',
                watchers_count: 123456,
            },
        ],
    };
    beforeEach(() => {
        nextRouter.useRouter = jest.fn();
        nextRouter.useRouter.mockImplementation(() => ({
            route: '/',
            pathName: '/repo',
            query: {repoId: '1234'},
        }));
    });

    it('should render without crashing', () => {
        useGithubData.mockReturnValue({});
        const container = render(<RepoDetailPage />);

        expect(container.baseElement).toMatchSnapshot();
    });

    it('should render loader', () => {
        useGithubData.mockReturnValue({loading: true});
        const container = render(<RepoDetailPage />);

        expect(container.baseElement).toMatchSnapshot();
    });

    it('should return a detail', () => {
        useGithubData.mockReturnValue({data: mockData});
        const container = render(<RepoDetailPage />);

        expect(container.baseElement).toMatchSnapshot();
    });
});
