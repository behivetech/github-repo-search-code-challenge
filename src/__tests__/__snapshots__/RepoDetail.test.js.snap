// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoDetailPage should render loader 1`] = `
<body>
  <div>
    <div
      class="repo-detail-page layout"
    >
      <header
        class="layout__header header"
      >
        <div
          class="header__left"
        >
          <h1
            class="headline--h1 headline"
          >
            <a
              class="link--no-underline link--on-primary link"
              href="/"
            >
              <i
                class="icon fa-icon fa-github fab"
              />
               Search
            </a>
          </h1>
        </div>
        <div
          class="header__right"
        >
           
          <a
            class="link--no-underline link--on-primary header__link link"
            href="/"
          >
            &lt;&lt; back to search
          </a>
        </div>
      </header>
      <main
        class="layout__main main-container"
      >
        <div
          class="loading"
        >
          <div
            class="loading__spinner"
          >
            <i
              class="rmwc-icon rmwc-icon--ligature material-icons loading-spinner icon"
            >
              synch
            </i>
          </div>
        </div>
      </main>
      <div
        aria-atomic="true"
        aria-hidden="true"
        aria-live="assertive"
        class="snackbar mdc-snackbar"
      >
        <div
          class="mdc-snackbar__surface"
        >
          <div
            aria-live="polite"
            class="mdc-snackbar__label"
            role="status"
          >
            <div
              class="notification-snackbar"
            />
            <div
              style="display: none;"
            />
          </div>
          <div
            class="mdc-snackbar__actions"
          >
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              code
            </button>
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RepoDetailPage should render without crashing 1`] = `
<body>
  <div>
    <div
      class="repo-detail-page layout"
    >
      <header
        class="layout__header header"
      >
        <div
          class="header__left"
        >
          <h1
            class="headline--h1 headline"
          >
            <a
              class="link--no-underline link--on-primary link"
              href="/"
            >
              <i
                class="icon fa-icon fa-github fab"
              />
               Search
            </a>
          </h1>
        </div>
        <div
          class="header__right"
        >
           
          <a
            class="link--no-underline link--on-primary header__link link"
            href="/"
          >
            &lt;&lt; back to search
          </a>
        </div>
      </header>
      <main
        class="layout__main main-container"
      >
        <section
          class="section--centered section--padding repo-detail section"
        >
          <div>
            Item not found.
          </div>
        </section>
      </main>
      <div
        aria-atomic="true"
        aria-hidden="true"
        aria-live="assertive"
        class="snackbar mdc-snackbar"
      >
        <div
          class="mdc-snackbar__surface"
        >
          <div
            aria-live="polite"
            class="mdc-snackbar__label"
            role="status"
          >
            <div
              class="notification-snackbar"
            />
            <div
              style="display: none;"
            />
          </div>
          <div
            class="mdc-snackbar__actions"
          >
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              code
            </button>
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RepoDetailPage should return a detail 1`] = `
<body>
  <div>
    <div
      class="repo-detail-page layout"
    >
      <header
        class="layout__header header"
      >
        <div
          class="header__left"
        >
          <h1
            class="headline--h1 headline"
          >
            <a
              class="link--no-underline link--on-primary link"
              href="/"
            >
              <i
                class="icon fa-icon fa-github fab"
              />
               Search
            </a>
          </h1>
        </div>
        <div
          class="header__right"
        >
           
          <a
            class="link--no-underline link--on-primary header__link link"
            href="/"
          >
            &lt;&lt; back to search
          </a>
        </div>
      </header>
      <main
        class="layout__main main-container"
      >
        <section
          class="section--centered section--padding repo-detail section"
        >
          <div
            class="repo-detail__header"
          >
            <img
              alt="Repo Avitar"
              class="repo-detail__avitar avitar"
              src="mock_avatar_url"
            />
            <div>
              <h2
                class="headline--h2 headline"
              >
                mock_name
              </h2>
              mock_login
            </div>
          </div>
          <dl
            class="repo-detail__dl"
          >
            <dt>
              Stargazers:
            </dt>
            <dd>
              123.5k
            </dd>
            <dt>
              Open Issues:
            </dt>
            <dd>
              5
            </dd>
            <dt>
              Open Issues:
            </dt>
            <dd>
              123
            </dd>
            <dt>
              Language:
            </dt>
            <dd>
              mock_language
            </dd>
            <dt>
              Created:
            </dt>
            <dd>
              01/01/2000 05:00:00 AM
            </dd>
            <dt>
              Updated:
            </dt>
            <dd>
              01/01/2000 05:00:00 AM
            </dd>
            <dt>
              Last Pushed:
            </dt>
            <dd>
              01/01/2000 05:00:00 AM
            </dd>
            <dt>
              License
            </dt>
            <dd>
              <a
                href="mock_url"
                rel="noreferrer"
                target="_blank"
              >
                mock_name
              </a>
            </dd>
          </dl>
          <div>
            mock_description
          </div>
        </section>
      </main>
      <div
        aria-atomic="true"
        aria-hidden="true"
        aria-live="assertive"
        class="snackbar mdc-snackbar"
      >
        <div
          class="mdc-snackbar__surface"
        >
          <div
            aria-live="polite"
            class="mdc-snackbar__label"
            role="status"
          >
            <div
              class="notification-snackbar"
            />
            <div
              style="display: none;"
            />
          </div>
          <div
            class="mdc-snackbar__actions"
          >
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              code
            </button>
            <button
              aria-hidden="true"
              class="rmwc-icon material-icons icon-button icon-button mdc-icon-button"
              data-mdc-ripple-is-unbounded="true"
              role="button"
              tabindex="0"
            >
              close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
